### PR TITLE
Minima method - stage 03

### PR DESCRIPTION
**Pull request by araihC**
Improved minima method for trigger detection: local maxima are also identified. These are required not to be too close (less MIN_PEAK_DISTANCE parameter, set by the user) and with sufficiently high intensity (compared to the maximum amplitude of the signal). Only one minimum can be detected within two accepted maxima.

### CODE REVIEW

File: pipeline/stage03_trigger_detection/scripts/minima.py [hunk 1]
```diff
@@ -1,29 +1,53 @@
...
-def detect_minima(asig, order, interpolation_points, interpolation):
+def detect_minima(asig, order, interpolation_points, interpolation, threshold_fraction,  Min_Peak_Distance):
+        
     signal = asig.as_array()
+    times = asig.times
     sampling_time = asig.times[1] - asig.times[0]
-
-    t_idx, channel_idx = argrelmin(signal, order=order, axis=0)
-
+    min_idx, channel_idx_minima = argrelmin(signal, order=order, axis=0)
```
Comment (rgutzen) [inline pipeline/stage03_trigger_detection/scripts/minima.py:15]: using all minima with `argrelmin` for the algorithm results in an unfavorable minima selection:
find_minima_example

using instead `find_peaks(-channel_signal, height=threshold[channel])` yields better results. 
However, then there is no mechanism for selection all minima... using separate thresholds for minima and peaks would work, but maybe there is a more elegant solution

**Review by rgutzen — CHANGES_REQUESTED** [review]

File: pipeline/stage03_trigger_detection/configs/config_template.yaml [hunk 2]
```diff
@@ -60,14 +60,17 @@ MOMENT_FACTOR: 3
 # Phase at which to define the upward transition [-pi, 0]
 TRANSITION_PHASE: -1.570796
 
-# BLOCK - Minima
```
[the remark this answers is not in the document]
Comment (araihC) [reply pipeline/stage03_trigger_detection/configs/config_template.yaml]: did not meant to

File: pipeline/stage03_trigger_detection/scripts/minima.py [hunk 3]
```diff
@@ -1,29 +1,53 @@
...
+    
+    # compute local minima times.
     if interpolation:
-
-        fitted_idx_times = np.zeros([len(t_idx)])
-
-        start_arr = t_idx - int(interpolation_points/2)
+        # parabolic fit around the local minima
+        fitted_idx_times = np.zeros([len(min_time_idx)])
+        start_arr = min_time_idx - 1 #int(interpolation_points/2)
```
[the remark this answers is not in the document]
Comment (araihC) [reply pipeline/stage03_trigger_detection/scripts/minima.py]: Only the right branch of the parabola should be fixed. On the left of transition times, there might be a plateau altering the fitting result.

File: pipeline/stage03_trigger_detection/scripts/minima.py (hunk 1 above)
Comment (araihC) [reply pipeline/stage03_trigger_detection/scripts/minima.py:15]: We should definitely think about it.

File: pipeline/stage03_trigger_detection/scripts/minima.py [hunk 4]
```diff
@@ -1,29 +1,54 @@
...
-def detect_minima(asig, order, interpolation_points, interpolation):
+def detect_minima(asig, order, interpolation_points, interpolation, threshold_fraction,  min_peak_distance):
+        
     signal = asig.as_array()
-    sampling_time = asig.times[1] - asig.times[0]
-
-    t_idx, channel_idx = argrelmin(signal, order=order, axis=0)
-
+    times = asig.times
+    sampling_rate = asig.sampling_rate
```
Comment (rgutzen) [inline pipeline/stage03_trigger_detection/scripts/minima.py:14]: `asig.sampling_rate.rescale('Hz').magnitude` to make sure it is multiplied correctly with `min_peak_distance` which is in s

File: pipeline/stage03_trigger_detection/scripts/minima.py [hunk 5]
```diff
@@ -1,29 +1,54 @@
...
 import quantities as pq
-from scipy.signal import argrelmin
+from scipy.signal import find_peaks
 import argparse
 from distutils.util import strtobool
 from utils import load_neo, write_neo, remove_annotations
 
 
-def detect_minima(asig, order, interpolation_points, interpolation):
+def detect_minima(asig, order, interpolation_points, interpolation, threshold_fraction,  min_peak_distance):
```
Comment (rgutzen) [inline pipeline/stage03_trigger_detection/scripts/minima.py:10]: the `order` parameter is now not longer used

File: pipeline/stage03_trigger_detection/scripts/minima.py [hunk 6]
```diff
@@ -1,29 +1,54 @@
...
-    sampling_time = asig.times[1] - asig.times[0]
-
-    t_idx, channel_idx = argrelmin(signal, order=order, axis=0)
-
+    times = asig.times
+    sampling_rate = asig.sampling_rate
+
+    amplitude_span = np.max(signal, axis=0) - np.min(signal, axis=0)
+    threshold = np.min(signal, axis=0) + threshold_fraction*(amplitude_span)
+    threshold_minima = np.max(-signal, axis=0) - threshold_fraction*(amplitude_span)
```
Comment (rgutzen) [inline pipeline/stage03_trigger_detection/scripts/minima.py:18]: I think the most flexible solution to control which minima to include is to have separate `maxima_threshold_fraction` and `minima_threshold_fraction` (basically replacing the `order` parameter). Or do you have another suggestion that we could explore?

File: pipeline/stage03_trigger_detection/scripts/minima.py [hunk 7]
```diff
@@ -1,29 +1,54 @@
...
+        min_time_idx = np.append(min_time_idx, clean_mins)
+        channel_idx = np.append(channel_idx, np.ones(len(clean_mins), dtype='int32')*channel)
+        
+    # compute local minima times.
     if interpolation:
-
-        fitted_idx_times = np.zeros([len(t_idx)])
-
-        start_arr = t_idx - int(interpolation_points/2)
+        # parabolic fit around the local minima
```
Comment (rgutzen) [inline pipeline/stage03_trigger_detection/scripts/minima.py:42]: add to comment that only the right side is used for the parabolic fit

**Review by rgutzen — APPROVED** [review]
Looks good to me now, I will merge it and test it in the method comparison

**Pull request merged**